python/tests/datasources/test_bjs_incarceration.py

```diff
@@ -0,0 +1,381 @@
+        check_like=True)
+
+
+def test_swap_race_col_names_to_codes():
+
+    _fake_df = pd.DataFrame({
+        std_col.STATE_NAME_COL: ["U.S. total", "Maine", "Florida", ],
+        'American Indian/Alaska Native': [1, 2, 3],
+        'Total': [4, 5, 6]
+    })
+
+    _expected_df_swapped_cols = pd.DataFrame({
+        std_col.STATE_NAME_COL: ["U.S. total", "Maine", "Florida", ],
+        'AIAN_NH': [1, 2, 3],
+        'ALL': [4, 5, 6]
+    })
+
+    assert_frame_equal(
+        swap_race_col_names_to_codes(_fake_df),
+        _expected_df_swapped_cols,
+        check_like=True)
+
+
+# MOCKS FOR READING IN TABLES
+
+
+def get_test_table_files():
+
+    loaded_tables = {}
+    for file in bjs_prisoners_tables.keys():
+        if file in bjs_prisoners_tables:
+            source_df = pd.read_csv(os.path.join(
+                TEST_DIR, f'bjs_test_input_{file}'),
+                encoding="ISO-8859-1",
+                thousands=',',
+                engine="python",
+            )
+
+            source_df = strip_footnote_refs_from_df(source_df)
+            source_df = missing_data_to_none(source_df)
+            loaded_tables[file] = set_state_col(source_df)
+
+    return loaded_tables
+
+
+def _get_pop_as_df(*args):
+    # retrieve fake ACS table subsets
+    data_source, table_name, _dtypes = args
+
+    df = pd.read_json(os.path.join(TEST_DIR, data_source,
+                                   f'{table_name}.json'),
+                      dtype={'state_fips': str})
+
+    return df
+
+
+# INTEGRATION TEST SETUP
+# Current working directory.
+THIS_DIR = os.path.dirname(os.path.abspath(__file__))
+TEST_DIR = os.path.join(THIS_DIR, os.pardir, "data", "bjs_corrections")
+
+GOLDEN_DATA = {
+    'race_and_ethnicity_national': os.path.join(TEST_DIR, 'bjs_test_output_race_and_ethnicity_national.json'),
+    'race_national': os.path.join(TEST_DIR, 'bjs_test_output_race_and_ethnicity_national.json'),
+    'race_and_ethnicity_state': os.path.join(TEST_DIR, 'bjs_test_output_race_and_ethnicity_state.json'),
+    'race_state': os.path.join(TEST_DIR, 'bjs_test_output_race_and_ethnicity_state.json'),
+    'age_national': os.path.join(TEST_DIR, 'bjs_test_output_age_national.json'),
+    'age_state': os.path.join(TEST_DIR, 'bjs_test_output_age_state.json'),
+    'sex_national': os.path.join(TEST_DIR, 'bjs_test_output_sex_national.json'),
+    'sex_state': os.path.join(TEST_DIR, 'bjs_test_output_sex_state.json'),
+}
+
+
+# RUN INTEGRATION TESTS ON NATIONAL LEVEL
+
+
+@ mock.patch('ingestion.gcs_to_bq_util.load_df_from_bigquery', side_effect=_get_pop_as_df)
+@ mock.patch('ingestion.gcs_to_bq_util.load_public_dataset_from_bigquery_as_df',
+             return_value=get_state_fips_codes_as_df())
+@ mock.patch('datasources.bjs.load_tables',
+             return_value=get_test_table_files())
+@ mock.patch('ingestion.gcs_to_bq_util.add_df_to_bq',
+             return_value=None)
+def testWriteNationalLevelToBq(mock_bq: mock.MagicMock,
```

Comment:
Maybe since my national_age and state_age are distinct from the other gen_breakdown fn I'll test those three fns, then add a test for the post_process() and finally use the existing test for the overall network calls etc

Reply:
re-added the `generate_breakdown_df` method to contain my gen_raw fns combined with the post_process() functions. Then in the tests, I test each distinct age breakdown fn, and run the sex/race one for one each at state and national. I left in the original test (without comparing the output) to confirm the network calls to BQ etc

Reply:
Thats awesome, thanks